rep-2004.rst

```diff
@@ -0,0 +1,681 @@
+Such a package may upgrade to 'Level 1' at a later date, if it proves to be a solution that people want to use in their products.
+
+Tools, like ``rostopic``\ , generally do not fall into this category either, but are not categorically excluded.
+For example, it may be the case that the tool which launches and verifies a ROS graph (``ros2launch``\ ) may need to be considered 'Level 1' for use in production systems.
+
+Package Requirements
+~~~~~~~~~~~~~~~~~~~~
+
+Requirements to be considered a 'Level 1' package:
+
+#. **Version Policy:**
+
+   #. Must have a version policy (e.g. ``semver``)
+   #. Must be at a stable version (e.g. for ``semver`` that means version >= 1.0.0)
+   #. Must have a strictly declared public API
+   #. Must have a policy for API stability
+   #. Must have a policy for ABI stability
+   #. Must have a policy that keeps API and ABI stability within a released ROS distribution
+
+#. **Change Control Process:**
+
+   #. Must have all code changes occur through a change request (e.g. pull request, merge request, etc.)
+   #. Must have confirmation of contributor origin (e.g. `DCO  <https://developercertificate.org/>`_, CLA, etc.)
+   #. Must have peer review policy for all change requests (e.g. require one or more reviewers)
+   #. Must have Continuous Integration (CI) policy for all change requests
+   #. Must have documentation policy for all change requests
+
+#. **Documentation:**
+
+   #. Must have documentation for each "feature" (e.g. for ``rclcpp``: create a node, publish a message, spin, etc.)
+   #. Must have documentation for each item in the public API (e.g. functions, classes, etc.)
+   #. Must have a declared license or set of licenses
+   #. Must have a copyright statement in each source file
+   #. Must have a "quality declaration" document, which declares the quality level and justifies how the package meets each of the requirements
+
+      #. *Must have a section in the repository's ``README`` which contains the "quality declaration" or links to it*
+      #. *Must register with a centralized list of 'Level N' packages, if one exists, to allow for peer review of the claim*
+
+#. **Testing:**
+
+   #. Must have system tests which cover all items in the "feature" documentation
+   #. Must have system, integration, and/or unit tests which cover all of the public API
+   #. Code Coverage:
+
+      #. *Must have code coverage tracking for the package*
+      #. *Must have and enforce a code coverage policy for new changes*
+
+   #. Performance:
+
+      #. *Must have performance tests (exceptions allowed if they don't make sense to have)*
+      #. *Must have a performance regression policy (i.e. blocking either changes or releases on unexpected performance regressions)*
+
+   #. Linters and Static Analysis:
+
+      #. *Must have a code style and enforce it*
```

Comment:
Propose to have an acceptance criteria of NO static analysis warnings remaining at this level

Reply:
We discussed this too, I think that we avoided saying that because there are some static code analysis tools which give false positives due to their aggressive nature. It was also not clear if suppressing these was sufficient to meet a "no failures/warnings" criteria or if it was even helpful to do that.

Also, any recommendations on what "no warnings" means for specific tools is, in my opinion, to prescriptive for this REP. It depends heavily on your target audience and your technology (which programming language you're using for example). As we discussed here:

https://github.com/ros-infrastructure/rep/pull/218#discussion_r381630440

I think the REP remaining broad in this case is good, simply because there's so much diversity under this point in terms of tools and outputs. It would be good to have this kind of stronger recommendations in the ROS 2 developer guide or as part of `ament_lint_common`. 

Do others have opinions about this?